File: Tests/RedlineCoreTests/IngestTests.swift
// Incremental reading of append only transcripts. The failure modes worth pinning are all
// about position: a line read twice doubles a day, a line skipped loses one, and a record
// parsed from half a line is worse than either because it looks like data.
import XCTest
@testable import RedlineCore

final class TranscriptTailTests: XCTestCase {
    private var dir: URL!

    override func setUpWithError() throws {
        dir = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("redline-tail-\(UUID().uuidString)")
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: dir)
    }

    private func write(_ text: String, to name: String = "t.jsonl") -> URL {
        let url = dir.appendingPathComponent(name)
        try? text.write(to: url, atomically: true, encoding: .utf8)
        return url
    }

    private func append(_ text: String, to url: URL) {
        guard let handle = try? FileHandle(forWritingTo: url) else { return }
        defer { try? handle.close() }
        _ = try? handle.seekToEnd()
        try? handle.write(contentsOf: Data(text.utf8))
    }

    func testReadsEveryCompleteLineWithItsOffset() {
        let url = write("one\ntwo\nthree\n")
        var seen: [(String, Int)] = []
        let next = TranscriptTail.read(url: url, from: 0) { seen.append(($0, $1)) }
        XCTAssertEqual(seen.map(\.0), ["one", "two", "three"])
        XCTAssertEqual(seen.map(\.1), [0, 4, 8])
        XCTAssertEqual(next, 14)
    }

    func testResumesWhereItStopped() {
        let url = write("one\ntwo\n")
        let first = TranscriptTail.read(url: url, from: 0) { _, _ in }
        append("three\n", to: url)
        var seen: [String] = []
        let next = TranscriptTail.read(url: url, from: first) { line, _ in seen.append(line) }
        XCTAssertEqual(seen, ["three"], "already read lines must not be read again")
        XCTAssertEqual(next, 14)
    }

    func testAPartialTrailingLineIsLeftForNextTime() {
        // A transcript being written right now ends mid record. Half a JSON object parsed
        // once is a bug that hides until the day it matters.
        let url = write("one\ntwo\nthr")
        var seen: [String] = []
        let next = TranscriptTail.read(url: url, from: 0) { line, _ in seen.append(line) }
        XCTAssertEqual(seen, ["one", "two"])
        XCTAssertEqual(next, 8, "the offset must stop at the last complete line")

        append("ee\n", to: url)
        var rest: [String] = []
        TranscriptTail.read(url: url, from: next) { line, _ in rest.append(line) }
        XCTAssertEqual(rest, ["three"], "the line arrives whole on the next pass")
    }

    func testALineLongerThanAChunkSurvives() {
        let long = String(repeating: "x", count: TranscriptTail.chunkSize + 1024)
        let url = write("short\n\(long)\nafter\n")
        var seen: [String] = []
        TranscriptTail.read(url: url, from: 0) { line, _ in seen.append(line) }
        XCTAssertEqual(seen.count, 3)
        XCTAssertEqual(seen[1].count, long.count)
        XCTAssertEqual(seen[2], "after")
    }

    func testATruncatedFileIsReadFromTheStart() {
        // Truncation and rewriting both mean the offsets no longer point where we think.
        let mark = IngestMark(path: "/tmp/x", provider: "Claude", size: 900,
                              byteOffset: 880, mtime: Date())
        XCTAssertEqual(TranscriptTail.startOffset(mark: mark, size: 400), 0)
        XCTAssertEqual(TranscriptTail.startOffset(mark: mark, size: 1200), 880)
        XCTAssertEqual(TranscriptTail.startOffset(mark: nil, size: 1200), 0)
    }

    func testBlankLinesAreSkippedWithoutLosingPosition() {
        let url = write("one\n\ntwo\n")
        var seen: [(String, Int)] = []
        let next = TranscriptTail.read(url: url, from: 0) { seen.append(($0, $1)) }
        XCTAssertEqual(seen.map(\.0), ["one", "two"])
        XCTAssertEqual(seen.map(\.1), [0, 5])
        XCTAssertEqual(next, 9)
    }
}

final class ClaudeIngestTests: XCTestCase {
    private var home: URL!
    private var projects: URL!
    private var warehouse: Warehouse!
    private var store: UsageStore!

    override func setUpWithError() throws {
        home = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("redline-ingest-\(UUID().uuidString)")
        projects = home.appendingPathComponent("projects/demo")
        try FileManager.default.createDirectory(at: projects, withIntermediateDirectories: true)
        warehouse = Warehouse(root: home.appendingPathComponent("history"))
        store = UsageStore(root: home.appendingPathComponent("projects"))
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: home)
    }

    /// One Claude Code assistant line, in the shape the parser actually reads.
    static func line(id: String, ts: String, input: Int = 100, output: Int = 10,
                     model: String = "claude-sonnet-5") -> String {
        """
        {"timestamp":"\(ts)","requestId":"req_\(id)","message":{"id":"\(id)",\
        "model":"\(model)","usage":{"input_tokens":\(input),"output_tokens":\(output),\
        "cache_read_input_tokens":0}}}
        """
    }

    private func writeTranscript(_ lines: [String], name: String = "a.jsonl") -> URL {
        let url = projects.appendingPathComponent(name)
        try? (lines.joined(separator: "\n") + "\n").write(to: url, atomically: true,
                                                          encoding: .utf8)
        return url
    }

    private func append(_ lines: [String], to url: URL) {
        guard let handle = try? FileHandle(forWritingTo: url) else { return }
        defer { try? handle.close() }
        _ = try? handle.seekToEnd()
        try? handle.write(contentsOf: Data((lines.joined(separator: "\n") + "\n").utf8))
    }

    func testFirstPassReadsEverythingAndSecondPassReadsNothing() {
        _ = writeTranscript([Self.line(id: "a", ts: "2026-08-18T10:00:00.000Z"),
                             Self.line(id: "b", ts: "2026-08-18T10:05:00.000Z")])
        XCTAssertEqual(store.ingest(into: warehouse), 2)
        XCTAssertEqual(store.ingest(into: warehouse), 0,
                       "an unchanged transcript costs a stat and nothing else")
        XCTAssertEqual(warehouse.entryCount, 2)
    }

    func testOnlyNewLinesAreIngested() {
        let url = writeTranscript([Self.line(id: "a", ts: "2026-08-18T10:00:00.000Z")])
        XCTAssertEqual(store.ingest(into: warehouse), 1)
        append([Self.line(id: "b", ts: "2026-08-18T11:00:00.000Z")], to: url)
        XCTAssertEqual(store.ingest(into: warehouse), 1)
        XCTAssertEqual(warehouse.entryCount, 2)
    }

    func testAResumedSessionCopiedIntoASecondFileIsCountedOnce() {
        // Resumed sessions copy identical message ids between transcripts. The id is the
        // dedup key, so the copy is recognised even though its byte position differs.
        let line = Self.line(id: "shared", ts: "2026-08-18T10:00:00.000Z")
        _ = writeTranscript([line], name: "one.jsonl")
        _ = writeTranscript([line], name: "two.jsonl")
        XCTAssertEqual(store.ingest(into: warehouse), 1)
        XCTAssertEqual(warehouse.entryCount, 1)
    }

    func testUsageSurvivesTheTranscriptBeingDeleted() {
        // The whole reason the store exists: Claude Code prunes its own projects directory.
        let url = writeTranscript([Self.line(id: "a", ts: "2026-08-18T10:00:00.000Z",
                                             input: 4321)])
        store.ingest(into: warehouse)
        warehouse.merge(entries: warehouse.entries(), config: Config())
        try? FileManager.default.removeItem(at: url)
        XCTAssertEqual(store.ingest(into: warehouse), 0)
        XCTAssertEqual(warehouse.entries().first?.input, 4321)
        XCTAssertNil(warehouse.ingestMark(path: url.path),
                     "the mark for a pruned transcript is dropped")
    }

    func testATranscriptRewrittenSmallerIsReadAgainFromTheStart() {
        let url = writeTranscript([Self.line(id: "a", ts: "2026-08-18T10:00:00.000Z"),
                                   Self.line(id: "b", ts: "2026-08-18T10:05:00.000Z")])
        store.ingest(into: warehouse)
        // Rewritten with different content, shorter than what was already read
        try? Data(Self.line(id: "c", ts: "2026-08-18T12:00:00.000Z").utf8 + [0x0A])
            .write(to: url)
        XCTAssertEqual(store.ingest(into: warehouse), 1)
        XCTAssertEqual(warehouse.entryCount, 3, "the earlier records are still held")
    }

    func testSyntheticModelsAndZeroUsageAreIgnored() {
        _ = writeTranscript([
            Self.line(id: "a", ts: "2026-08-18T10:00:00.000Z", model: "<synthetic>"),
            Self.line(id: "b", ts: "2026-08-18T10:01:00.000Z", input: 0, output: 0),
            Self.line(id: "c", ts: "2026-08-18T10:02:00.000Z"),
        ])
        XCTAssertEqual(store.ingest(into: warehouse), 1)
    }

    func testTailedAndWholeFileParsesAgree() {
        let lines = (0..<20).map {
            Self.line(id: "m\($0)", ts: "2026-08-18T10:\(String(format: "%02d", $0)):00.000Z")
        }
        _ = writeTranscript(lines)
        store.ingest(into: warehouse)
        let scanned = store.scan(lookbackDays: 3650,
                                 now: Date(timeIntervalSince1970: 1_787_000_000))
        XCTAssertEqual(warehouse.entryCount, scanned.count)
        XCTAssertEqual(warehouse.entries().reduce(0) { $0 + $1.input },
                       scanned.reduce(0) { $0 + $1.input })
    }
}

File: windows/tests/Redline.Core.Tests/IngestTests.cs
// Incremental reading of append only transcripts. A line read twice doubles a day, a line
// skipped loses one, and a record parsed from half a line looks like data.
using System.Text;
using Redline.Core;

namespace Redline.Core.Tests;

public sealed class TranscriptTailTests : IDisposable
{
    private readonly string _dir;

    public TranscriptTailTests()
    {
        _dir = Path.Combine(Path.GetTempPath(), "redline-tail-" + Guid.NewGuid());
        Directory.CreateDirectory(_dir);
    }

    public void Dispose()
    {
        try { Directory.Delete(_dir, true); } catch { }
    }

    private string Write(string text, string name = "t.jsonl")
    {
        var path = Path.Combine(_dir, name);
        File.WriteAllText(path, text, new UTF8Encoding(false));
        return path;
    }

    private static void Append(string text, string path) =>
        File.AppendAllText(path, text, new UTF8Encoding(false));

    [Fact]
    public void ReadsEveryCompleteLineWithItsOffset()
    {
        var path = Write("one\ntwo\nthree\n");
        var seen = new List<(string, long)>();
        var next = TranscriptTail.Read(path, 0, (l, o) => seen.Add((l, o)));
        Assert.Equal(new[] { "one", "two", "three" }, seen.Select(s => s.Item1));
        Assert.Equal(new long[] { 0, 4, 8 }, seen.Select(s => s.Item2));
        Assert.Equal(14, next);
    }

    [Fact]
    public void ResumesWhereItStopped()
    {
        var path = Write("one\ntwo\n");
        var first = TranscriptTail.Read(path, 0, (_, _) => { });
        Append("three\n", path);
        var seen = new List<string>();
        var next = TranscriptTail.Read(path, first, (l, _) => seen.Add(l));
        Assert.Equal(new[] { "three" }, seen); // already read lines must not be read again
        Assert.Equal(14, next);
    }

    [Fact]
    public void APartialTrailingLineIsLeftForNextTime()
    {
        // A transcript being written right now ends mid record. Half a JSON object parsed
        // once is a bug that hides until the day it matters.
        var path = Write("one\ntwo\nthr");
        var seen = new List<string>();
        var next = TranscriptTail.Read(path, 0, (l, _) => seen.Add(l));
        Assert.Equal(new[] { "one", "two" }, seen);
        Assert.Equal(8, next); // the offset must stop at the last complete line

        Append("ee\n", path);
        var rest = new List<string>();
        TranscriptTail.Read(path, next, (l, _) => rest.Add(l));
        Assert.Equal(new[] { "three" }, rest); // the line arrives whole on the next pass
    }

    [Fact]
    public void ALineLongerThanAChunkSurvives()
    {
        var longLine = new string('x', TranscriptTail.ChunkSize + 1024);
        var path = Write($"short\n{longLine}\nafter\n");
        var seen = new List<string>();
        TranscriptTail.Read(path, 0, (l, _) => seen.Add(l));
        Assert.Equal(3, seen.Count);
        Assert.Equal(longLine.Length, seen[1].Length);
        Assert.Equal("after", seen[2]);
    }

    [Fact]
    public void ATruncatedFileIsReadFromTheStart()
    {
        // Truncation and rewriting both mean the offsets no longer point where we think.
        var mark = new IngestMark("/tmp/x", "Claude", 900, 880, DateTimeOffset.UtcNow);
        Assert.Equal(0, TranscriptTail.StartOffset(mark, 400));
        Assert.Equal(880, TranscriptTail.StartOffset(mark, 1200));
        Assert.Equal(0, TranscriptTail.StartOffset(null, 1200));
    }

    [Fact]
    public void BlankLinesAreSkippedWithoutLosingPosition()
    {
        var path = Write("one\n\ntwo\n");
        var seen = new List<(string, long)>();
        var next = TranscriptTail.Read(path, 0, (l, o) => seen.Add((l, o)));
        Assert.Equal(new[] { "one", "two" }, seen.Select(s => s.Item1));
        Assert.Equal(new long[] { 0, 5 }, seen.Select(s => s.Item2));
        Assert.Equal(9, next);
    }
}

public sealed class ClaudeIngestTests : IDisposable
{
    private readonly string _home;
    private readonly string _projects;
    private readonly Warehouse _warehouse;
    private readonly UsageStore _store;

    public ClaudeIngestTests()
    {
        _home = Path.Combine(Path.GetTempPath(), "redline-ingest-" + Guid.NewGuid());
        _projects = Path.Combine(_home, "projects", "demo");
        Directory.CreateDirectory(_projects);
        _warehouse = new Warehouse(Path.Combine(_home, "history"));
        _store = new UsageStore(Path.Combine(_home, "projects"));
    }

    public void Dispose()
    {
        _warehouse.Dispose();
        try { Directory.Delete(_home, true); } catch { }
    }

    /// One Claude Code assistant line, in the shape the parser actually reads.
    internal static string Line(string id, string ts, int input = 100, int output = 10,
                                string model = "claude-sonnet-5") =>
        $"{{\"timestamp\":\"{ts}\",\"requestId\":\"req_{id}\",\"message\":{{\"id\":\"{id}\"," +
        $"\"model\":\"{model}\",\"usage\":{{\"input_tokens\":{input},\"output_tokens\":{output}," +
        "\"cache_read_input_tokens\":0}}}";

    private string WriteTranscript(IEnumerable<string> lines, string name = "a.jsonl")
    {
        var path = Path.Combine(_projects, name);
        File.WriteAllText(path, string.Join("\n", lines) + "\n", new UTF8Encoding(false));
        return path;
    }

    private static void Append(IEnumerable<string> lines, string path) =>
        File.AppendAllText(path, string.Join("\n", lines) + "\n", new UTF8Encoding(false));

    [Fact]
    public void FirstPassReadsEverythingAndSecondPassReadsNothing()
    {
        WriteTranscript(new[] { Line("a", "2026-08-18T10:00:00.000Z"),
                                Line("b", "2026-08-18T10:05:00.000Z") });
        Assert.Equal(2, _store.Ingest(_warehouse));
        Assert.Equal(0, _store.Ingest(_warehouse)); // an unchanged transcript costs a stat and nothing else
        Assert.Equal(2, _warehouse.EntryCount);
    }

    [Fact]
    public void OnlyNewLinesAreIngested()
    {
        var path = WriteTranscript(new[] { Line("a", "2026-08-18T10:00:00.000Z") });
        Assert.Equal(1, _store.Ingest(_warehouse));
        Append(new[] { Line("b", "2026-08-18T11:00:00.000Z") }, path);
        Assert.Equal(1, _store.Ingest(_warehouse));
        Assert.Equal(2, _warehouse.EntryCount);
    }

    [Fact]
    public void AResumedSessionCopiedIntoASecondFileIsCountedOnce()
    {
        // Resumed sessions copy identical message ids between transcripts. The id is the
        // dedup key, so the copy is recognised even though its byte position differs.
        var line = Line("shared", "2026-08-18T10:00:00.000Z");
        WriteTranscript(new[] { line }, "one.jsonl");
        WriteTranscript(new[] { line }, "two.jsonl");
        Assert.Equal(1, _store.Ingest(_warehouse));
        Assert.Equal(1, _warehouse.EntryCount);
    }

    [Fact]
    public void UsageSurvivesTheTranscriptBeingDeleted()
    {
        // The whole reason the store exists: Claude Code prunes its own projects directory.
        var path = WriteTranscript(new[] { Line("a", "2026-08-18T10:00:00.000Z", input: 4321) });
        _store.Ingest(_warehouse);
        _warehouse.Merge(_warehouse.Entries(), new Config());
        File.Delete(path);
        Assert.Equal(0, _store.Ingest(_warehouse));
        Assert.Equal(4321, _warehouse.Entries().FirstOrDefault()?.Input);
        Assert.Null(_warehouse.IngestMark(path)); // the mark for a pruned transcript is dropped
    }

    [Fact]
    public void ATranscriptRewrittenSmallerIsReadAgainFromTheStart()
    {
        var path = WriteTranscript(new[] { Line("a", "2026-08-18T10:00:00.000Z"),
                                           Line("b", "2026-08-18T10:05:00.000Z") });
        _store.Ingest(_warehouse);
        // Rewritten with different content, shorter than what was already read
        File.WriteAllBytes(path, Encoding.UTF8.GetBytes(Line("c", "2026-08-18T12:00:00.000Z") + "\n"));
        Assert.Equal(1, _store.Ingest(_warehouse));
        Assert.Equal(3, _warehouse.EntryCount); // the earlier records are still held
    }

    [Fact]
    public void SyntheticModelsAndZeroUsageAreIgnored()
    {
        WriteTranscript(new[]
        {
            Line("a", "2026-08-18T10:00:00.000Z", model: "<synthetic>"),
            Line("b", "2026-08-18T10:01:00.000Z", input: 0, output: 0),
            Line("c", "2026-08-18T10:02:00.000Z"),
        });
        Assert.Equal(1, _store.Ingest(_warehouse));
    }

    [Fact]
    public void TailedAndWholeFileParsesAgree()
    {
        var lines = Enumerable.Range(0, 20)
            .Select(i => Line($"m{i}", $"2026-08-18T10:{i:00}:00.000Z")).ToList();
        WriteTranscript(lines);
        _store.Ingest(_warehouse);
        var scanned = _store.Scan(3650, Database.FromSeconds(1_787_000_000));
        Assert.Equal(scanned.Count, _warehouse.EntryCount);
        Assert.Equal(scanned.Sum(e => e.Input), _warehouse.Entries().Sum(e => e.Input));
    }
}
